common/types/unknown_test.go

```diff
@@ -22,43 +22,239 @@ import (
 )
 
 func TestIsUnknown(t *testing.T) {
-	if IsUnknown(Unknown{}) != true {
+	if IsUnknown(&Unknown{}) != true {
 		t.Error("IsUnknown(Unknown{}) returned false, wanted true")
 	}
 	if IsUnknown(Bool(true)) != false {
 		t.Error("IsUnknown(true) returned true, wanted false")
 	}
 }
 
+func TestNewAttribute(t *testing.T) {
+	if NewAttributeTrail("") != unspecifiedAttribute {
+		t.Error("An empty attribute must be the unspecified attribute")
+	}
+	if NewAttributeTrail("v").Equal(unspecifiedAttribute) {
+		t.Error("A non-empty attribute must not be equal to the unspecified attribute")
+	}
+}
+
+func TestAttributeEquals(t *testing.T) {
+	tests := []struct {
+		a     *AttributeTrail
+		b     *AttributeTrail
+		equal bool
+	}{
+		{
+			a:     unspecifiedAttribute,
+			b:     NewAttributeTrail(""),
+			equal: true,
+		},
+		{
+			a:     NewAttributeTrail("a"),
+			b:     NewAttributeTrail(""),
+			equal: false,
+		},
+		{
+			a:     NewAttributeTrail("a"),
+			b:     NewAttributeTrail("a"),
+			equal: true,
+		},
+		{
+			a:     QualifyAttribute[string](NewAttributeTrail("a"), "b"),
+			b:     NewAttributeTrail("a"),
+			equal: false,
+		},
+		{
+			a:     QualifyAttribute[string](NewAttributeTrail("a"), "b"),
+			b:     QualifyAttribute[int64](NewAttributeTrail("a"), 1),
```

Comment:
could you add a test case for a[1] == a[1u]? I'm not sure whether these should be treated as equal or not since maps treat them as equivalent with heterogeneous numeric equality